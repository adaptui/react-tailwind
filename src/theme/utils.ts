import { defaults, isString, isUndefined, mergeWith, tcm } from "../utils";

import {
  DefaultTheme,
  ExtendableDefaultTheme,
  PartialDefaultTheme,
} from "./types";

export const extendTheme = <T = ExtendableDefaultTheme>(theme: T): T => {
  return theme;
};

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
function collectExtends(items: ExtendableDefaultTheme[]) {
  return items.reduce((merged, { extend }) => {
    return mergeWith(merged, extend, (mergedValue, extendValue) => {
      if (isUndefined(mergedValue)) {
        return [extendValue];
      }

      if (Array.isArray(mergedValue)) {
        return [extendValue, ...mergedValue];
      }

      return [extendValue, mergedValue];
    });
  }, {});
}

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
export function mergeThemes(themes: PartialDefaultTheme[]) {
  return {
    ...themes.reduce<DefaultTheme>(
      (merged, theme) => defaults(merged, theme),
      {} as DefaultTheme,
    ),

    // In order to resolve n config objects, we combine all of their `extend` properties
    // into arrays instead of objects so they aren't overridden.
    extend: collectExtends(themes),
  };
}

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
export function mergeExtensions({
  extend,
  ...theme
}: DefaultTheme & { extend?: PartialDefaultTheme }) {
  return mergeWith(theme, extend, (themeValue, extendValue) => {
    return mergeWith(
      themeValue,
      ...extendValue,
      (merged: any, value: any[]) => {
        if (isString(merged) && isString(value)) {
          return tcm(merged, value);
        }

        return undefined;
      },
    );
  });
}
