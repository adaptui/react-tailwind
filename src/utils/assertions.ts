import { Dict } from "./types";

// Null / Undefined assertions
export function isUndefined(value: any): value is undefined {
  return typeof value === "undefined" || value === undefined;
}

export { isNull } from "@renderlesskit/react";

// String assertions
export function isString(value: any): value is string {
  return Object.prototype.toString.call(value) === "[object String]";
}

// Array assertions
export function isArray<T>(value: any): value is Array<T> {
  return Array.isArray(value);
}

// Object assertions
export const isObject = (value: any): value is Dict => {
  const type = typeof value;

  return value != null && type === "object" && !isArray(value);
};

// Function assertions
export function isFunction(value: any): value is Function {
  return typeof value === "function";
}

// SSR check
export function canUseDOM() {
  return !!(
    typeof window !== "undefined" &&
    window.document &&
    window.document.createElement
  );
}

export const isBrowser = canUseDOM();

export function isEmptyObject(value: any) {
  return isObject(value) && Object.keys(value).length === 0;
}
