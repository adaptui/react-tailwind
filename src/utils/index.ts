export * from "./assertions";
export * from "./createContext";
export * from "./debounce";
export * from "./global-types";
export * from "./object";
export * from "./react";
export * from "./tailwindMerge";
export * from "./types";
