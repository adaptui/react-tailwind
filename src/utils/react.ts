import * as React from "react";

import { isFunction } from "./assertions";
import { cx } from "./tailwindMerge";
import {
  As,
  ComponentWithAs,
  Dict,
  PropsWithAs,
  RenderPropType,
} from "./types";

/**
 * @template Props Component Props
 * @template RefProp HTML intrinsic type, eg: HTMLDivElement
 * @template DefaultType string, eg: "div"
 * @param component
 */
export function forwardRefWithAs<
  Props,
  RefProp = any,
  DefaultType extends As = any,
>(
  component: React.ForwardRefRenderFunction<
    RefProp,
    PropsWithAs<Props, DefaultType>
  >,
) {
  return React.forwardRef(component) as unknown as ComponentWithAs<
    Props,
    DefaultType
  >;
}

// From Chakra Utils
export function runIfFn<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  return isFunction(valueOrFn) ? valueOrFn(...args) : valueOrFn;
}

export function runIfFnChildren<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  // Know bug but this is the only way to make it work
  // {({ isChecked }) => {
  //   return (
  //     <>
  //       <SwitchText data-testid="testid-mode">
  //         {isChecked ? DARK_MODE : LIGHT_MODE}
  //       </SwitchText>
  //     </>
  //   );
  // }}
  // Need a wrapper to get the children when the children is a function
  // @ts-ignore
  return isFunction(valueOrFn) ? valueOrFn(...args).props.children : valueOrFn;
}

/**
 * Gets only the valid children of a component,
 * and ignores any nullish or falsy child.
 *
 * @param children the children
 */
export function getValidChildren(children: React.ReactNode) {
  return React.Children.toArray(children).filter(child =>
    React.isValidElement(child),
  ) as React.ReactElement[];
}

// Merge library & user prop
export const passProps = (icon: RenderPropType, props?: Dict) => {
  return React.isValidElement(icon)
    ? React.cloneElement(icon, {
        ...props,
        ...icon.props,
        className: cx(props?.className, icon.props.className),
      })
    : runIfFn(icon, { ...props });
};

// Add a11y to the icon passed
export const withIconA11y = (icon: RenderPropType, props?: Dict) => {
  return passProps(icon, {
    role: "img",
    focusable: false,
    "aria-hidden": true,
    ...props,
  });
};

export const getComponentProps = <T, P>(
  componentMaps: Dict<string>,
  children: T,
  props: P,
) => {
  const normalizedChildren = runIfFnChildren(children, props);
  const validChildren = getValidChildren(normalizedChildren);
  const componentProps: Dict = {};
  const finalChildren: React.ReactNode[] = [];

  validChildren.forEach(child => {
    // @ts-ignore
    if (componentMaps[child.type.displayName]) {
      componentProps[
        // @ts-ignore
        componentMaps[child.type.displayName]
      ] = child.props;
    } else {
      finalChildren.push(child);
    }
  });

  return { componentProps, finalChildren };
};
