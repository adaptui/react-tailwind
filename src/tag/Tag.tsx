import { ButtonHTMLProps, ButtonOptions, useButton } from "reakit";
import { createComponent, createHook } from "@renderlesskit/react";

import { CloseIcon } from "../icons";
import { useTheme } from "../theme";
import { cx, RenderPropType, tcm, withIconA11y } from "../utils";

import { TAG_KEYS } from "./__keys";

export type TagOptions = ButtonOptions & {
  /**
   * How large should the tag be?
   *
   * @default md
   */
  size?: keyof Renderlesskit.GetThemeValue<"tag", "size", "default">;

  /**
   * How the tag should look?
   *
   * @default solid
   */
  variant?: keyof Renderlesskit.GetThemeValue<"tag", "variant", "default">;

  /**
   * If added, the tag will show an icon before the tag's text.
   */
  prefix?: RenderPropType;

  /**
   * If added, the tag will allow to show an icon before the tag's text.
   */
  closable?: boolean;

  /**
   * If added, the tag will show an icon after the tag's text.
   */
  suffix?: RenderPropType;
};

export type TagHTMLProps = Omit<ButtonHTMLProps, "prefix"> & {};

export type TagProps = TagOptions & TagHTMLProps;

export const useTag = createHook<TagOptions, TagHTMLProps>({
  name: "Tag",
  compose: useButton,
  keys: TAG_KEYS,

  useOptions(options, htmlProps) {
    const { size = "md", variant = "solid", ...restOptions } = options;

    return { size, variant, ...restOptions };
  },

  useProps(options, htmlProps) {
    const {
      size = "md",
      variant = "solid",
      prefix,
      closable = false,
      suffix = closable ? <CloseIcon /> : null,
      disabled,
    } = options;

    let {
      className: htmlClassName,
      children: htmlChildren,
      ...restHtmlProps
    } = htmlProps;

    const theme = useTheme("tag");
    const className = cx(
      theme.base,
      theme.size.default[size],
      theme.variant.default[variant],
      disabled
        ? theme.variant.disabled[variant]
        : tcm(
            theme.variant.hover[variant],
            theme.variant.active[variant],
            theme.variant.focus[variant],
          ),
      htmlClassName,
    );
    const prefixStyles = cx(theme.size.prefix[size]);
    const suffixStyles = cx(theme.size.suffix[size]);

    const children = (
      <>
        {prefix ? (
          <>{withIconA11y(prefix, { className: prefixStyles })}</>
        ) : null}
        <span>{htmlChildren}</span>
        {closable && suffix ? (
          <>{withIconA11y(suffix, { className: suffixStyles })}</>
        ) : null}
      </>
    );

    return { className, children, ...restHtmlProps };
  },
});

export const Tag = createComponent({
  as: "button",
  memo: true,
  useHook: useTag,
});
