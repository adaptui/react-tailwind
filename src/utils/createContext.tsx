import * as React from "react";
import { isUndefined } from "lodash";

export interface CreateContextOptions<ContextType> {
  /**
   * The display name of the context
   */
  name?: string;

  /**
   * Default context value.
   */
  defaultContext?: ContextType;

  /**
   * If `true`, React will throw if context is `null` or `undefined`
   * In some cases, you might want to support nested context, so you can set it to `false`
   */
  strict?: boolean;

  /**
   * Error message to throw if the context is `undefined`
   */
  errorMessage?: string;
}

/**
 * Creates a named context, provider, and hook.
 *
 * @param options create context options
 */
export function createContext<ContextType extends object | null>(
  options: CreateContextOptions<ContextType> = {},
) {
  const {
    name = "Provider",
    defaultContext,
    strict = true,
    errorMessage = "Seems you forgot to wrap `useContext` within the Context Provider",
  } = options;

  const Context = React.createContext<ContextType | undefined>(defaultContext);

  function Provider(props: ContextType & { children: React.ReactNode }) {
    const { children, ...context } = props;

    // Idea to memoize within the createContext from [radix ui](https://github.com/radix-ui/primitives/blob/cefcb6d9281cf203b0ab54b49f85725dc203df85/packages/react/context/src/createContext.tsx#L13)
    // Only re-memoize when prop values change
    const value = React.useMemo(
      () => context,
      // eslint-disable-next-line react-hooks/exhaustive-deps
      Object.values(context),
    ) as ContextType;

    return <Context.Provider value={value}>{children}</Context.Provider>;
  }

  Provider.displayName = name;

  function useContext() {
    const context = React.useContext(Context);

    if (context) return context;

    if (!isUndefined(defaultContext)) return defaultContext;

    if (strict) {
      throw new Error(errorMessage);
    }

    return undefined;
  }

  return [Provider, useContext] as const;
}
