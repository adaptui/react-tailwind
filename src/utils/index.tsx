import * as React from "react";
import { createTailwindMerge } from "tailwind-merge";

import { cx } from "../utils";

import {
  As,
  ComponentWithAs,
  Dict,
  PropsWithAs,
  RenderPropType,
} from "./types";

export interface CreateContextOptions {
  /**
   * If `true`, React will throw if context is `null` or `undefined`
   * In some cases, you might want to support nested context, so you can set it to `false`
   */
  strict?: boolean;
  /**
   * Error message to throw if the context is `undefined`
   */
  errorMessage?: string;
  /**
   * The display name of the context
   */
  name?: string;
}

export type CreateContextReturn<T> = [
  React.Provider<T>,
  () => T,
  React.Context<T>,
];

/**
 * Creates a named context, provider, and hook.
 *
 * @param options create context options
 */
export function createContext<ContextType>(options: CreateContextOptions = {}) {
  const {
    strict = true,
    errorMessage = "useContext: `context` is undefined. Seems you forgot to wrap component within the Provider",
    name,
  } = options;

  const Context = React.createContext<ContextType | undefined>(undefined);

  Context.displayName = name;

  function useContext() {
    const context = React.useContext(Context);

    if (!context && strict) {
      throw new Error(errorMessage);
    }

    return context;
  }

  return [
    Context.Provider,
    useContext,
    Context,
  ] as CreateContextReturn<ContextType>;
}

// Function assertions
export function isFunction(value: any): value is Function {
  return typeof value === "function";
}

// From Chakra Utils
export function runIfFn<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  return isFunction(valueOrFn) ? valueOrFn(...args) : valueOrFn;
}

export function runIfFnChildren<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  // Know bug but this is the only way to make it work
  // {({ isChecked }) => {
  //   return (
  //     <>
  //       <SwitchText data-testid="testid-mode">
  //         {isChecked ? DARK_MODE : LIGHT_MODE}
  //       </SwitchText>
  //     </>
  //   );
  // }}
  // Need a wrapper to get the children when the children is a function
  // @ts-ignore
  return isFunction(valueOrFn) ? valueOrFn(...args).props.children : valueOrFn;
}

/**
 * Gets only the valid children of a component,
 * and ignores any nullish or falsy child.
 *
 * @param children the children
 */
export function getValidChildren(children: React.ReactNode) {
  return React.Children.toArray(children).filter(child =>
    React.isValidElement(child),
  ) as React.ReactElement[];
}

// SSR check
export function canUseDOM() {
  return !!(
    typeof window !== "undefined" &&
    window.document &&
    window.document.createElement
  );
}

export const isBrowser = canUseDOM();

// Array assertions
export function isArray<T>(value: any): value is Array<T> {
  return Array.isArray(value);
}

// Object assertions
export const isObject = (value: any): value is Dict => {
  const type = typeof value;

  return value != null && type === "object" && !isArray(value);
};

export const objectKeys = <T extends Dict>(obj: T) =>
  Object.keys(obj) as unknown as (keyof T)[];

// Corresponds to 10 frames at 60 Hz.
// A few bytes payload overhead when lodash/debounce is ~3 kB and debounce ~300 B.
export const debounce = (func: Function, wait = 166) => {
  let timeout: any;
  function debounced(this: any, ...args: any) {
    // eslint-disable-next-line consistent-this
    const that = this;
    const later = () => {
      func.apply(that, args);
    };
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
  }

  debounced.clear = () => {
    clearTimeout(timeout);
  };

  return debounced;
};

// String assertions
export function isString(value: any): value is string {
  return Object.prototype.toString.call(value) === "[object String]";
}

// Merge library & user prop
export const passProps = (icon: RenderPropType, props?: Dict) => {
  return React.isValidElement(icon)
    ? React.cloneElement(icon, {
        ...props,
        ...icon.props,
        className: cx(props?.className, icon.props.className),
      })
    : runIfFn(icon, { ...props });
};

// Add a11y to the icon passed
export const withIconA11y = (icon: RenderPropType, props?: Dict) => {
  return passProps(icon, {
    role: "img",
    focusable: false,
    "aria-hidden": true,
    ...props,
  });
};

export function isUndefined(value: any): value is undefined {
  return typeof value === "undefined" || value === undefined;
}

export const tcm = createTailwindMerge(getDefaultConfig => {
  const defaultConfig = getDefaultConfig();

  return {
    ...defaultConfig,
    classGroups: {
      ...defaultConfig.classGroups,
      "font-size": [
        ...defaultConfig.classGroups["font-size"],
        // Defined custom text presets
        {
          text: ["cxs", "paragraph-cxs", "paragraph-sm", "2base"],
        },
      ],
    },
  };
});

export const getComponentProps = <T, P>(
  componentMaps: Dict<string>,
  children: T,
  props: P,
) => {
  const normalizedChildren = runIfFnChildren(children, props);
  const validChildren = getValidChildren(normalizedChildren);
  const componentProps: Dict = {};
  const finalChildren: React.ReactNode[] = [];

  validChildren.forEach(child => {
    // @ts-ignore
    if (componentMaps[child.type.displayName]) {
      componentProps[
        // @ts-ignore
        componentMaps[child.type.displayName]
      ] = child.props;
    } else {
      finalChildren.push(child);
    }
  });

  return { componentProps, finalChildren };
};

/**
 * @template Props Component Props
 * @template RefProp HTML intrinsic type, eg: HTMLDivElement
 * @template DefaultType string, eg: "div"
 * @param component
 */
export function forwardRefWithAs<
  Props,
  RefProp = any,
  DefaultType extends As = any,
>(
  component: React.ForwardRefRenderFunction<
    RefProp,
    PropsWithAs<Props, DefaultType>
  >,
) {
  return React.forwardRef(component) as unknown as ComponentWithAs<
    Props,
    DefaultType
  >;
}

export * from "./types";
export { cx } from "@renderlesskit/react";
