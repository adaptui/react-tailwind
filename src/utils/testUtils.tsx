import {
  render as RtlRender,
  RenderOptions,
  waitFor,
  RenderResult,
} from "@testing-library/react";
import * as React from "react";
import { RunOptions } from "axe-core";
import { configureAxe } from "jest-axe";

// @ts-ignore
import tailwindConfig from "../../tailwind.config";
import { RenderlesskitProvider } from "../theme";
export * from "@testing-library/react";

type Render = (
  children: React.ReactNode,
  options?: RenderOptions,
) => RenderResult;

export const render: Render = (children, options = {}) => {
  return RtlRender(
    <RenderlesskitProvider tailwindConfig={tailwindConfig}>
      {children}
    </RenderlesskitProvider>,
    options,
  );
};

export const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

type UI = Parameters<typeof render>[0] | Element;
type TestA11YOptions = RenderOptions & {
  axeOptions?: RunOptions;
};

export const testA11y = async (
  ui: UI,
  { axeOptions, ...options }: TestA11YOptions = {},
) => {
  jest.useRealTimers();

  const container = React.isValidElement(ui)
    ? render(ui, options).container
    : ui;

  // @ts-ignore
  const results = await axe(container, axeOptions);

  expect(results).toHaveNoViolations();
};

const DELAY = 0;
const LOAD_IMAGE = "load.png";
const ERROR_IMAGE = "error.png";
const orignalImage = window.Image;

export const mockImage = (loadState: string) => {
  jest.useFakeTimers();

  (window.Image as unknown) = class MockImage {
    onload: () => void = () => {};
    onerror: () => void = () => {};
    src: string = "";
    constructor() {
      if (loadState === LOAD_IMAGE) {
        setTimeout(() => {
          this.onload();
        }, DELAY);
      }
      if (loadState === ERROR_IMAGE) {
        setTimeout(() => {
          this.onerror();
        }, DELAY);
      }
      return this;
    }
  };
};

mockImage.DELAY = DELAY;
mockImage.LOAD_IMAGE = LOAD_IMAGE;
mockImage.ERROR_IMAGE = ERROR_IMAGE;
mockImage.ERROR_IMAGE = ERROR_IMAGE;
mockImage.load = () => mockImage(LOAD_IMAGE);
mockImage.error = () => mockImage(ERROR_IMAGE);
mockImage.restoreMock = () => (window.Image = orignalImage);
mockImage.advanceTimer = async () => {
  return await waitFor(() => {
    jest.advanceTimersByTime(mockImage.DELAY);
  });
};
