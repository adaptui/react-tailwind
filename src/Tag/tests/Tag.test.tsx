import * as React from "react";
import { fireEvent, render, screen } from "@testing-library/react";

import { Tag } from "../Tag";
import { testA11y } from "../../utils/testUtils";

const spyWarn = jest.spyOn(console, "warn");

beforeEach(() => {
  spyWarn.mockReset();
});

afterEach(() => {
  expect(spyWarn).not.toHaveBeenCalled();
});

describe("<Tag />", () => {
  it("should be closable", () => {
    const onCloseFn = jest.fn();
    render(
      <Tag closable onClose={onCloseFn}>
        hello
      </Tag>,
    );

    fireEvent.click(screen.getByRole("button"));
    expect(onCloseFn).toHaveBeenCalled();
  });

  it("close icon should not be visible if closable=false", () => {
    render(<Tag>hello</Tag>);

    expect(
      screen.queryByTestId("testid-close-element"),
    ).not.toBeInTheDocument();
  });

  it("custom suffix", () => {
    render(
      <Tag closable suffix={<p>Suffix</p>}>
        hello
      </Tag>,
    );

    expect(screen.queryByTestId("testid-close-element")).toBeInTheDocument();
    expect(screen.getByText(/suffix/i)).toBeInTheDocument();
  });

  // TODO: Fix spy warn not firing
  it.skip("should throw warning if suffix is provided with and closable is false", () => {
    render(<Tag suffix={<p>Suffix</p>}>hello</Tag>);

    expect(
      screen.queryByTestId("testid-close-element"),
    ).not.toBeInTheDocument();

    expect(spyWarn).toBeCalledTimes(1);
  });

  it("should not have a11y violations", async () => {
    await testA11y(<Tag>Ally</Tag>);
  });
});
