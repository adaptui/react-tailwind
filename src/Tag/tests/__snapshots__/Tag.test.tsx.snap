// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag /> should render Tag 1`] = `
<DocumentFragment>
  <span
    class="font-sans font-semibold bg-gray-100 text-gray-800 inline-flex items-center justify-center appearance-none rounded-md transition-all relative whitespace-nowrap align-middle outline-none w-auto select-none disabled:cursor-not-allowed disabled:opacity-40 h-10 min-w-10 text-base px-4"
  >
    hello
  </span>
</DocumentFragment>
`;

exports[`<Tag /> should render Tag with prefix & suffix 1`] = `
<DocumentFragment>
  <span
    class="font-sans font-semibold bg-gray-100 text-gray-800 inline-flex items-center justify-center appearance-none rounded-md transition-all relative whitespace-nowrap align-middle outline-none w-auto select-none disabled:cursor-not-allowed disabled:opacity-40 h-10 min-w-10 text-base px-4"
  >
    <span
      class="flex mr-2"
    >
      <p>
        prefix
      </p>
    </span>
    hello
    <button
      class="flex ml-2"
      data-testid="testid-close-element"
      id="id-ksdndd"
      tabindex="0"
    >
      <p>
        suffix
      </p>
    </button>
  </span>
</DocumentFragment>
`;
